HW1/P4/P4.py

```diff
@@ -0,0 +1,37 @@
+import findspark
+findspark.init()
+from pyspark import SparkContext, SparkConf
+
+conf = SparkConf().setAppName('P4').setMaster('local')
+sc = SparkContext(conf=conf)
+
+# Graph and BFS to determine the number of touched nodes of a signle-sorce search
+
+def helper(x, y):
+	if y[0] != y[1]:
+		return (x, y)
+
+def getGraph(source):
+	source_p = sc.textFile(source)
+	comic_char = source_p.map(lambda line: (line.split('"' or ',')[3], line.split('"' or ',')[1]))
+	comic_two_char = comic_char.join(comic_char)
+	comic_two_char_distinct = comic_two_char.filter(lambda (x, y): helper(x, y))
+	two_char_distinct = comic_two_char_distinct.map(lambda (x, y): y)
+	return two_char_distinct
+
+graph = getGraph('./source.csv')
+
+def BFS(start):
+	RDD_iter = sc.parallelize([(start, 0)])
+	RDD_visited = RDD_iter
+	while True:
+		RDD_iter = graph.join(RDD_iter).distinct().map(lambda (x, y): (y[0], 0)).distinct().subtract(RDD_visited)
```

Comment:
I think some documentation/comments would be helpful here. Also, I guess join() & distinct() would cause a shuffle at this point since graph and RDD_iter are not copartitioned (if you use partitionBy with the same # of partitions then assuming the two RDDs have the same keys, they should end up with same partitioner). I'm not sure about union() but I think it also causes a shuffle if things are not copartitioned.